src/DataStreams/ColumnGathererStream.cpp

```diff
@@ -111,4 +114,89 @@ void ColumnGathererStream::readSuffixImpl()
             profile_info.rows / seconds, ReadableSize(profile_info.bytes / seconds));
 }
 
+void ColumnGathererStream::gatherLowCardinality(ColumnLowCardinality &column_res)
+{
+    if (!enable_low_cardinality_merge_new_algo)
+    {
+        gather(column_res);
+        return ;
+    }
+
+    row_sources_buf.nextIfAtEnd();
+    RowSourcePart * row_source_pos = reinterpret_cast<RowSourcePart *>(row_sources_buf.position());
+    RowSourcePart * row_sources_end = reinterpret_cast<RowSourcePart *>(row_sources_buf.buffer().end());
+
+    size_t cur_block_preferred_size = static_cast<size_t>(row_sources_end - row_source_pos);
+    column_res.reserve(cur_block_preferred_size);
+
+    if (cardinalityDict)
+        column_res.loadDictionaryFrom(*cardinalityDict);
+
+    while (row_source_pos < row_sources_end)
+    {
+        RowSourcePart row_source = *row_source_pos;
+        size_t source_num = row_source.getSourceNum();
+        Source & source = sources[source_num];
+        bool source_skip = row_source.getSkipFlag();
+        ++row_source_pos;
+
+        if (source.pos >= source.size) /// Fetch new block from source_num part
+        {
+            fetchNewBlock(source, source_num);
+            column_res.mergeGatherColumn(*source.column, source.index_map);
+            if (!source.index_map.empty())
+            {
+                auto *low_ref = const_cast<IColumn *>(source.column);
+                auto *low_src = typeid_cast<ColumnLowCardinality *>(low_ref);
+                low_src->transformIndex(source.index_map, column_res.getDictionary().size());
+            }
+        }
+
+        size_t len = 1;
+        size_t max_len = std::min(static_cast<size_t>(row_sources_end - row_source_pos), source.size - source.pos); // interval should be in the same block
+        while (len < max_len && row_source_pos->data == row_source.data)
+        {
+            ++len;
+            ++row_source_pos;
+        }
+
+        row_sources_buf.position() = reinterpret_cast<char *>(row_source_pos);
+
+        if (!source_skip)
+        {
+            if (len == 1)
+                column_res.insertIndexFrom(*source.column, source.pos);
+            else
+                column_res.insertIndexRangeFrom(*source.column, source.pos, len);
+        }
+
+        source.pos += len;
+    }
+
+    bool need_save_dict = false;
+    for (auto &ref : sources)
+    {
+        if (ref.pos >= ref.size)
+        {
+            ref.index_map.clear();
+        }
+        else
+        {
+            need_save_dict = true;
+        }
+    }
```

Comment:
I think we need a better condition to drop dictionary.
Now, we drop it only if all columns are at the end, but I suspect it really almost never happens.

I suppose, we should decide to drop index (and rebuild a new one) if it's size is too big. Probably, bigger then `max(sum(source[i].dictionary_size) * 2, block_preferred_size, settings.low_cardinality_max_dictionary_size)`. (In this case we will also need to update sources)

Reply:
That depends, from my test the first and second level merge rarely need to save dict.
Here only the dictionary may cost more memory, and the index map bounded by granule size. If we do the drop may save a little memory but may cause performance degrade. Since the dictionary size should be smaller than the block_preferred_size